openml/flows/flow.py

```diff
@@ -1,80 +1,358 @@
+
+            param_dict['oml:default_value'] = self.parameters[key]
+            if self.parameters_meta_info[key]['description'] is not None:
+                param_dict['oml:description'] = self.parameters_meta_info[key].\
+                    get('description')
+
+            for key_, value in param_dict.items():
+                if key_ is not None and not isinstance(key_, six.string_types):
+                    raise ValueError('Parameter name %s cannot be serialized '
+                                     'because it is of type %s. Only strings '
+                                     'can be serialized.' % (key_, type(key_)))
+                if value is not None and not isinstance(value, six.string_types):
+                    raise ValueError('Parameter value %s cannot be serialized '
+                                     'because it is of type %s. Only strings '
+                                     'can be serialized.' % (value, type(value)))
+
             flow_parameters.append(param_dict)
 
         flow_dict['oml:flow']['oml:parameter'] = flow_parameters
 
-        flow_xml = xmltodict.unparse(flow_dict, pretty=True)
+        components = []
+        for key in self.components:
+            component_dict = OrderedDict()
+            component_dict['oml:identifier'] = key
+            component_dict['oml:flow'] = \
+                self.components[key].__to_dict()['oml:flow']
 
-        # A flow may not be uploaded with the encoding specification..
-        flow_xml = flow_xml.split('\n', 1)[-1]
-        return flow_xml
+            for key_ in component_dict:
+                # We only need to check if the key is a string, because the
+                # value is a flow. The flow itself is valid by recursion
+                if key_ is not None and not isinstance(key_, six.string_types):
+                    raise ValueError('Parameter name %s cannot be serialized '
+                                     'because it is of type %s. Only strings '
+                                     'can be serialized.' % (key_, type(key_)))
+
+            components.append(component_dict)
+
+        flow_dict['oml:flow']['oml:component'] = components
+
+        flow_dict['oml:flow']['oml:tag'] = self.tags
+
+        if self.binary_url is not None:
+            flow_dict['oml:flow']['oml:binary_url'] = self.binary_url
+        if self.binary_format is not None:
+            flow_dict['oml:flow']['oml:binary_format'] = self.binary_format
+        if self.binary_md5 is not None:
+            flow_dict['oml:flow']['oml:binary_md5'] = self.binary_md5
+
+        return flow_dict
+
+    @classmethod
+    def _from_xml(cls, xml_dict):
+        """Create a flow from an xml description.
+
+        Parameters
+        ----------
+        xml_dict : dict
+            Dictionary representation of the flow as created by _to_dict()
+
+        Returns
+        -------
+            OpenMLFlow
+
+        """
+        dic = xml_dict["oml:flow"]
```

Comment:
I will try to make the code more compact, but I'm afraid it won't be that simple. Two things which have to be done separately is to put the parameter information in two different dictionaries (one with the parameter value and one with meta information; allows easier access of the parameter values) and to check whether a component/parameter/tag is actually a single tag or multiple tags. xmltodict will return a dictionary in the first, and a list in the second case. I will add a brief comment on this to the code.